### PR TITLE
Sync #f9c82f48

### PR DESCRIPTION
**Pull request by veaba**
## Description of Problem

https://github.com/vuejs/docs/compare/6d4e543c...f9c82f48

### CODE REVIEW

File: src/guide/reusability/custom-directives.md [hunk 1]
```diff
@@ -84,24 +95,68 @@ export default {
...
 
-将一个自定义指令全局注册到应用层级也是一种常见的做法：
+It is also common to globally register custom directives at the app level:
 
 ```js
 const app = createApp({})
 
-// 使 v-focus 在所有组件中都可用
-app.directive('focus', {
+// make v-highlight usable in all components
```
Comment (KiritaniAyaka) [inline src/guide/reusability/custom-directives.md:107]:
```suggestion
// 使 v-highlight 在所有组件中都可用
```

File: src/guide/reusability/custom-directives.md [hunk 2]
```diff
@@ -84,24 +95,68 @@ export default {
...
+
+```vue-html
+<input v-focus />
+```
+
+</div>
+
+该指令比 `autofocus` 属性更有用，因为它不仅在页面加载时有效，而且在 Vue 动态插入元素时也有效！
+
+在可能的情况下，建议使用内置指令如 `v-bind` 进行声明式模板，因为它们更高效，对服务器渲染也更友好。
```
Comment (KiritaniAyaka) [inline src/guide/reusability/custom-directives.md:159]:
```suggestion
建议尽可能使用 `v-bind` 等内置指令声明模板，因为它们更高效，对服务端渲染也更友好。
```

File: src/guide/reusability/custom-directives.md [hunk 3]
```diff
@@ -84,24 +95,68 @@ export default {
...
 ```
 
-:::tip
-只有当所需功能只能通过直接的 DOM 操作来实现时，才应该使用自定义指令。其他情况下应该尽可能地使用 `v-bind` 这样的内置指令来声明式地使用模板，这样更高效，也对服务端渲染更友好。
-:::
+## When to use custom directives {#when-to-use}
+
+Custom directives should only be used when the desired functionality can only be achieved via direct DOM manipulation.
+
+A common example of this is a `v-focus` custom directive that brings an element into focus.
```
Comment (KiritaniAyaka) [inline src/guide/reusability/custom-directives.md:117]:
```suggestion
只有当所需功能只能通过直接的 DOM 操作来实现时，才应该使用自定义指令。

一个常见例子是使元素获取焦点的 `v-focus` 指令。
```

File: src/guide/scaling-up/testing.md [hunk 4]
```diff
@@ -261,13 +288,21 @@ Vitest 和基于浏览器的运行器之间的主要区别是速度和执行上
 
 ### 推荐方案 {#recommendation-2}
 
-- [Cypress](https://www.cypress.io/)
+- [Playwright](https://playwright.dev/) 也是一个非常好的端到端测试解决方案，支持所有的现代渲染引擎，包括 Chromium、WebKit 和 Firefox。在 Windows、Linux 和 macOS 上进行本地或 CI 测试、无头测试，或使用适用于 Android 和 Mobile Safari 的 Google Chrome 的原生移动端模拟测试。它拥有信息丰富的用户界面、出色的调试能力、内置断言、并行处理功能以及追踪功能，旨在消除不稳定的测试。它还提供对[组件测试](https://docs.cypress.io/guides/component-testing/introduction)的支持，目前实验阶段。Playwright 由微软开源维护。
```
Comment (KiritaniAyaka) [inline src/guide/scaling-up/testing.md:291]:
```suggestion
- [Playwright](https://playwright.dev/) 是一个非常好的端到端测试解决方案，支持 Chromium、WebKit 和 Firefox。在 Windows、Linux 和 macOS 上进行本地或 CI 测试、无头测试，或使用适用于 Android 和 Mobile Safari 的 Google Chrome 的原生移动端模拟测试。它拥有信息丰富的用户界面、出色的调试能力、内置断言、并行处理功能以及追踪功能，旨在消除不稳定的测试。它还提供对[组件测试](https://docs.cypress.io/guides/component-testing/introduction)的支持，但目前处于实验阶段。Playwright 由微软开源并维护。
```

File: src/guide/scaling-up/testing.md [hunk 5]
```diff
@@ -261,13 +288,21 @@ Vitest 和基于浏览器的运行器之间的主要区别是速度和执行上
...
 
-  总的来说，我们认为 Cypress 提供了最完整的端到端解决方案，其具有信息丰富的图形界面、出色的调试性、内置断言、存根、抗剥落性、并行化和快照等诸多特性。而且如上所述，它还提供对 [组件测试](https://docs.cypress.io/guides/component-testing/introduction) 的支持。它支持基于 Chromium 的浏览器、Firefox 和 Electron。但 WebKit 被标记为实验性支持。
+- [Cypress](https://www.cypress.io/) 具有信息丰富的图形界面、出色的调试性、内置断言、存根、抗剥落性、并行化和快照等诸多特性。而且如上所述，它还提供对 [组件测试](https://docs.cypress.io/guides/component-testing/introduction) 的支持。它支持基于 Chromium 的浏览器、Firefox 和 Electron。但 WebKit 被标记为实验性支持。Cypress 采用 MIT 许可，但某些功能，如并行化，需要订阅 Cypress Cloud。
```
Comment (KiritaniAyaka) [inline src/guide/scaling-up/testing.md:293]:
```suggestion
- [Cypress](https://www.cypress.io/) 具有信息丰富的图形界面、出色的调试性、内置断言、存根、抗剥落性、并行化和快照等诸多特性。而且如上所述，它还提供对 [组件测试](https://docs.cypress.io/guides/component-testing/introduction) 的支持。它支持基于 Chromium 的浏览器、Firefox 和 Electron。但 WebKit 被标记为实验性支持。Cypress 采用 MIT 许可，但并行化等部分功能需要订阅 Cypress Cloud。
```

**Review by KiritaniAyaka — CHANGES_REQUESTED** [review]

**Review by veaba — COMMENTED** [review]
apply the suggustions.

File: src/guide/extras/web-components.md [hunk 6]
```diff
@@ -236,31 +238,277 @@ export function register() {
...
-    Counter: typeof Counter
+  interface GlobalComponents {
+    // 请务必在此处输入 Vue 组件类型（SomeComponent，*而不是* SomeElement）。
+    // 自定义元素的名称中需要连字符，因此请在此处使用连字符元素名称。
+    'some-element': typeof SomeComponent
+  }
+}
+```
+
+## 非 Vue Web Components 和 TypeScript {#non-vue-web-components-and-typescript}
```
Comment (veaba) [inline src/guide/extras/web-components.md:303]: 待翻译

File: src/guide/reusability/custom-directives.md [hunk 7]
```diff
@@ -36,44 +49,42 @@ const vFocus = {
...
-  <input v-focus placeholder="This should be focused" />
+  <p v-highlight>This sentence is important!</p>
 </div>
 
-假设你还未点击页面中的其他地方，那么上面这个 input 元素应该会被自动聚焦。该指令比 `autofocus` attribute 更有用，因为它不仅仅可以在页面加载完成后生效，还可以在 Vue 动态插入元素后生效。
-
 <div class="composition-api">
 
-在 `<script setup>` 中，任何以 `v` 开头的驼峰式命名的变量都可以被用作一个自定义指令。在上面的例子中，`vFocus` 即可以在模板中以 `v-focus` 的形式使用。
+在 `<script setup>` 中，任何以 `v` 开头的驼峰式命名的变量都可以当作自定义指令使用。在上述例子中，`vHighlight` 可以在模板中以 `v-highlight`的形式使用。
```
Comment (Jinjiang) [inline src/guide/reusability/custom-directives.md:76]:
```suggestion
在 `<script setup>` 中，任何以 `v` 开头的驼峰式命名的变量都可以当作自定义指令使用。在上述例子中，`vHighlight` 可以在模板中以 `v-highlight` 的形式使用。
```

**Comment by veaba** [comment]
close #967
see https://github.com/vuejs-translations/docs-zh-cn/commit/7d926de67b7d9895f2b59c614133b83ace993af3

**Pull request merged**